### DISCUSSION TITLE
Pfsense OpenVPN parsing

### DISCUSSION THREAD

**Discussion by sleepingbel** [Security-Onion-Solutions/securityonion#3039] [Show and tell] [answered]
Hello All,

If you have a Pfsense router and you use the OpenVPN feature, you can use the following custum parser scripts in /opt/so/saltstack/local/salt/elasticsearch/files/ingest/

**Syslog**

```
 {
  "description" : "syslog",
  "processors" : [
    {
        "dissect": {
                "field": "message",
                "pattern" : "%{message}",
                "on_failure": [ { "drop" : { } } ]
        },
        "remove": {
                "field": [ "type", "agent" ],
                "ignore_failure": true
        }
    },
    {
        "grok":
                {
                        "field": "message",
                        "patterns": [
                                        "^<%{INT:syslog.priority}> %{DATA:source.application}: %{DATE_EU:syslog.timestamp} %{GREEDYDATA:event_data.message}$",
					"^<%{INT:syslog.priority}>%{DATA:syslog.timestamp} %{WORD:source.application}(\\[%{DATA:pid}\\])?: %{GREEDYDATA:real_message}$",
                                        "^%{SYSLOGTIMESTAMP:syslog.timestamp} %{SYSLOGHOST:syslog.host} %{SYSLOGPROG:syslog.program}: CEF:0\\|%{DATA:vendor}\\|%{DATA:product}\\|%{GREEDYDATA:message2}$"
                                    ],
                        "ignore_failure": true
                }
    },
    { "set": { "if": "ctx.source?.application == 'filterlog'", "field": "dataset", "value": "firewall", "ignore_failure": true  } },
    { "set": { "if": "ctx.source?.application == 'openvpn'", "field": "dataset", "value": "vpn", "ignore_failure": true  } },
    { "set": { "if": "ctx.vendor != null", "field": "module", "value": "{{ vendor }}", "ignore_failure": true } },
    { "set": { "if": "ctx.product != null", "field": "dataset", "value": "{{ product }}", "ignore_failure": true } },
    { "set": { "field": "ingest.timestamp",      "value": "{{ @timestamp }}"       } },
    { "date": { "if": "ctx.syslog?.timestamp != null", "field": "syslog.timestamp",       "target_field": "@timestamp",   "formats": ["MMM  d HH:mm:ss", "MMM dd HH:mm:ss", "ISO8601", "UNIX"], "ignore_failure": true  } },
    { "set": { "if": "ctx.source?.application == 'sshguard'", "field": "module", "value": "pfsense", "override": true  } },
    { "set": { "if": "ctx.source?.application == 'openvpn'", "field": "module", "value": "pfsense", "override": true  } },
    { "remove":         { "field": ["pid", "program"],  "ignore_missing": true, "ignore_failure": true  } },
    { "pipeline": { "if": "ctx.vendor != null && ctx.product != null", "name": "{{ vendor }}.{{ product }}", "ignore_failure": true } },
    { "pipeline": { "if": "ctx.dataset == 'firewall'", "name": "filterlog", "ignore_failure": true } },
    { "pipeline": { "if": "ctx.dataset == 'vpn'", "name": "vpn", "ignore_failure": true } },
    { "pipeline":       { "name": "common"                                             } }
  ]
}
```

**Comment by DE4DB4BE** [comment, marked as answer]
Hey there, thanks for this! I can't get the parsing to function properly, though. I'm not capturing raw logs on pfsense, not sure if that's required. Are there any other steps needed on the SO side to make this work? I did run `sudo so-elasticsearch-restart` and the parser loaded fine. Thanks!

**Reply by sleepingbel** [reply]
Do not forget to forward the vpn logs from your router with syslog. You need also to place a double backslash, this is in the original code but is hidden by gitlab in the view mode.
And yes, the Raw log output changed after a version update of openvpn this is the new parser:

``` 

{
  "description" : "vpn",
  "processors" : [
  {
    "dissect": {
      "field": "real_message",
      "pattern": "%{real_message}"
    }
  },
  {
    "grok": {
      "field": "real_message",
      "patterns": [
        "^\\user '%{DATA:user.name}' %{GREEDYDATA:event_data.message}$",
        "^%{IP:source.ip}:%{POSINT:source.port} %{DATA:event_data.process.name}: %{GREEDYDATA:event_data.message}$",
        "^%{IP:source.ip}:%{POSINT:source.port} (\\[%{DATA:user.name}\\]) %{GREEDYDATA:event_data.message}$",
	"^%{DATA:user.name}/%{IP:source.ip}:%{POSINT:source.port} \\[%{DATA:scrap}\\] %{GREEDYDATA:event_data.message}$",
	"^%{DATA:user.name}/%{IP:source.ip}:%{POSINT:source.port} %{DATA:event_data.process.name}: %{DATA:event_data.notice.message} \\IPv4=%{IP:client.ip}, %{GREEDYDATA:event_data.message}$",
        "^%{DATA:user.name}/%{IP:source.ip}:%{POSINT:source.port} %{DATA:event_data.process.name}: %{GREEDYDATA:event_data.message}$",
        "^%{DATA:event_data.process.name} '%{DATA:event_data.process.args}' %{DATA:scrap} '%{DATA:user.name}' %{DATA:scrap} '%{IP:source.ip}' - %{GREEDYDATA:event_data.message}$",
	"^%{DATA:event_data.process.name}: %{GREEDYDATA:event_data.notice.message}"
      ],
      "ignore_failure": true
    }
  },
  {    "set": {      "field": "_index",      "value": "so-vpn", "override": true    }  },
  {    "set": {      "field": "module",      "value": "pfsense", "override": true    }  },
  {    "set": {      "field": "dataset",      "value": "vpn", "override": true    }  },
  {    "remove": {      "field": ["real_message", "scrap"],      "ignore_failure": true    }  }
 ]
}

```

**Reply by DE4DB4BE** [reply]
Thanks for the updated parser. I added it and have enabled raw logging and the config loaded perfectly fine! Should I be able to see these logs in Kibana or Hunt via SOC?

**Reply by sleepingbel** [reply]
Yes in in Kibana by filtering on dataset:vpn and module:pfsense in the hunt tab dataset:vpn

**Reply by dougburks** [reply]
@sleepingbel May we include this in our default parsers?

**Reply by DE4DB4BE** [reply]
Weird, doesn't appear to be working. I tried with both VPN parsing configs and I only see firewall logs. I've attached a screenshots from Kibana and Pfsense.
<img width="630" alt="image" src="https://user-images.githubusercontent.com/15071240/160215110-6fe72515-69af-4cf7-9778-d6e16aad1983.png">
<img width="589" alt="image" src="https://user-images.githubusercontent.com/15071240/160215176-97c09877-5cdf-4338-aa2b-a2464f1d10e2.png">

**Reply by sleepingbel** [reply]
> @sleepingbel May we include this in our default parsers?

Yes Doug

**Reply by sleepingbel** [reply]
The parsing is working for me look

![afbeelding](https://user-images.githubusercontent.com/29328171/160229864-79a14c4c-2b59-47ee-bd99-830224c28c34.png)
![Knipsel](https://user-images.githubusercontent.com/29328171/160229968-397a97ba-3921-463b-886b-bb5c0b5d6b28.PNG)

You do need to make a VPN connection to see the logs, this parser only parse actual vpn connections.

**Reply by dougburks** [reply]
> > @sleepingbel May we include this in our default parsers?
> 
> Yes Doug

I've created https://github.com/Security-Onion-Solutions/securityonion/issues/7656 to add this.

**Reply by DE4DB4BE** [reply]
Interesting - I'm wondering if I'm placing the parser in the wrong location or doing some other thing wrong.
I've placed the parser here: `/opt/so/conf/elasticsearch/ingest` then ran `sudo so-elasticsearch-restart`
Am I missing another step for this to be adopted? Is there maybe a log where I can see if there are any errors loading the config?

Thank you so much for all the help with this, I really appreciate it!

**Reply by sleepingbel** [reply]
Did you placed the syslog and vpn parser also in  /opt/so/saltstack/local/salt/elasticsearch/files/ingest/?

The changes on the syslog parser forward the vpn logs to the vpn parser and the vpn parser ingest these in elastic

If you place them directly in /opt/so/conf/elasticsearch/ingest the sysmon parser gets overwritten by salt to the default parser

**Reply by DE4DB4BE** [reply]
There it is! It was because of me putting the parsers in the incorrect location. All is working perfectly, thanks again very much for your patience and assistance with everything!

**Comment by utkarshborawake** [comment]


**Reply by dougburks** [reply]
@utkarshborawake First, please review the discussion guidelines at https://github.com/Security-Onion-Solutions/securityonion/discussions/1720:
```
Start a new discussion instead of replying to somebody else's discussion
Please search to see if you can find similar discussions that may help you. However, in order to avoid confusion, please do NOT reply to somebody else's discussion with your own issue. Instead, please start a new discussion and in that new discussion you can provide a hyperlink to the related discussion.

Tagging Individuals
Please do not tag an individual in a discussion unless that individual has already volunteered to help you in that discussion.
```

Second, please try the Elastic integration for pfSense:
https://docs.securityonion.net/en/2.4/pfsense.html#elastic-integration-for-pfsense

If you have further questions or problems, please start a new discussion rather than replying to this old discussion.
